### PR TITLE
Add Thiel's U Statistic (Uncertainty) Metric

### LINKED ISSUE #1273: Add Uncertainty coefficient (Theil's U)
## 🚀 Feature

Add uncertainty coefficient

In [statistics](https://en.wikipedia.org/wiki/Statistics), the uncertainty coefficient, also called proficiency, entropy coefficient or Theil's U, is a measure of nominal [association](https://en.wikipedia.org/wiki/Association_(statistics)). (Source: https://en.wikipedia.org/wiki/Uncertainty_coefficient)

### Motivation

Add correlation metrics specific to categorical variables

### Additional context

Nice blog post on this: https://towardsdatascience.com/the-search-for-categorical-correlation-a1cf7f1888c9

See #1298

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?
Implements `Theil's U` statistic for nominal (categorical) data.
Fixes #1273 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/functional/nominal/thiels_u.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    # handle the nan in data
+    preds, target = _handle_nan_in_data(preds, target, nan_strategy, nan_replace_value)
+
+    # compute conditional entropy
+    s_xy = conditional_entropy(preds, target)
+
+    x_counter = Counter(preds.numpy())
+    total_occurrences = sum(x_counter.values())
+    p_x = list(map(lambda n: n / total_occurrences, x_counter.values()))
+    s_x = ss.entropy(p_x)
```
Comment (stancld) [inline src/torchmetrics/functional/nominal/thiels_u.py:136]: Believe we should be able to implement this without a dependency on `scipy` :]

File: src/torchmetrics/nominal/thiels_u.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.nominal.thiels_u import _thiels_u_input_validation, theils_u
+from torchmetrics.metric import Metric
+
+
+class Thiels_U(Metric):
```
Comment (stancld) [inline src/torchmetrics/nominal/thiels_u.py:24]:
```suggestion
class ThielsU(Metric):
```

File: src/torchmetrics/functional/nominal/thiels_u.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+def _thiels_u_input_validation(nan_strategy: str, nan_replace_value: Optional[Union[int, float]]) -> None:
+    if nan_strategy not in ["replace", "drop"]:
+        raise ValueError(
+            f"Argument `nan_strategy` is expected to be one of `['replace', 'drop']`, but got {nan_strategy}"
+        )
+    if nan_strategy == "replace" and not isinstance(nan_replace_value, (int, float)):
+        raise ValueError(
+            "Argument `nan_replace` is expected to be of a type `int` or `float` when `nan_strategy = 'replace`, "
+            f"but got {nan_replace_value}"
+        )
```
Comment (stancld) [inline src/torchmetrics/functional/nominal/thiels_u.py:36]: Pls rebase on `master` once #1334 is merged. I'll refactor this method into the common nominal utils.

File: src/torchmetrics/functional/nominal/thiels_u.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    y_counter = Counter(y)
+    xy_counter = Counter(list(zip(x, y)))
+    total_occurrences = sum(y_counter.values())
+    entropy = 0.0
+    for xy in xy_counter.keys():
+        p_xy = xy_counter[xy] / total_occurrences
+        p_y = y_counter[xy[1]] / total_occurrences
+        entropy += p_xy * math.log(p_y / p_xy, log_base)
+    return torch.tensor(entropy)
```
Comment (stancld) [inline src/torchmetrics/functional/nominal/thiels_u.py:88]: Would it be possible this part vectorize with `torch`? Not sure at the first glance, but maybe it's doable 🤔

File: src/torchmetrics/functional/nominal/thiels_u.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def theils_u(
+    preds: Tensor,
+    target: Tensor,
+    nan_strategy: Literal["replace", "drop"] = "replace",
+    nan_replace_value: Optional[Union[int, float]] = 0.0,
+    _PRECISION: float = 1e-13,
```
Comment (stancld) [inline src/torchmetrics/functional/nominal/thiels_u.py:96]: I'd omit this parameter used in the reference library and be in align with the Cramer's V implementation.

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 1 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: I updated it without any package dependence. Can you check it out?

File: src/torchmetrics/nominal/thiels_u.py (hunk 2 above)
Comment (shenoynikhil) [reply src/torchmetrics/nominal/thiels_u.py]: Updated it.

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: Fixed this.

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 1 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: Yeah, fixed this, now using the tensor based `confmat` based solution

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 3 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: Done.

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 4 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: Fixed it using the confmat based solution used in crammer's v implementation.

File: src/torchmetrics/functional/nominal/thiels_u.py (hunk 5 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/nominal/thiels_u.py]: Done.

File: src/torchmetrics/functional/nominal/theils_u.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    s_xy = _conditional_entropy_compute(confmat)
+
+    # compute H(x)
+    total_occurrences = confmat.sum()
+    p_x = confmat.sum(0) / total_occurrences
+    s_x = -torch.sum(p_x * torch.log(p_x))
+
+    # compute u statistic
+    if s_x == torch.Tensor(0):
+        return torch.Tensor(1.0)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/nominal/theils_u.py:75]:
```suggestion
    if s_x == 0:
        return torch.tensor(0, device=confmat.device)
```

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Comment by SkafteNicki** [comment]
@stancld please re-review :]

File: src/torchmetrics/functional/__init__.py [hunk 7]
```diff
@@ -47,6 +47,7 @@
     pearsons_contingency_coefficient,
     pearsons_contingency_coefficient_matrix,
 )
+from torchmetrics.functional.nominal.theils_u import theils_u
```
Comment (stancld) [inline src/torchmetrics/functional/__init__.py:50]:
```suggestion
from torchmetrics.functional.nominal.theils_u import theils_u, theils_u_matrix
```

File: src/torchmetrics/functional/__init__.py [hunk 8]
```diff
@@ -171,6 +172,7 @@
     "structural_similarity_index_measure",
     "stat_scores",
     "symmetric_mean_absolute_percentage_error",
+    "theils_u",
     "total_variation",
```
Comment (stancld) [inline src/torchmetrics/functional/__init__.py:176]:
```suggestion
    "theils_u",
    "theils_u_matrix",
    "total_variation",
```

File: src/torchmetrics/functional/nominal/theils_u.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        nan_strategy: Indication of whether to replace or drop ``NaN`` values
+        nan_replace_value: Value to replace ``NaN`s when ``nan_strategy = 'replace```
+
+    Returns:
+        Non-reduced confusion matrix
+    """
+    preds = preds.argmax(1) if preds.ndim == 2 else preds
+    target = target.argmax(1) if target.ndim == 2 else target
+    preds, target = _handle_nan_in_data(preds, target, nan_strategy, nan_replace_value)
+    return _multiclass_confusion_matrix_update(preds, target, num_classes)
```
Comment (stancld) [inline src/torchmetrics/functional/nominal/theils_u.py:102]: Just for consistency, place `update` in front of `compute` in the code.

**Review by stancld — APPROVED** [review]
LGTM! Very nice, thanks for the contribution!

File: tests/unittests/nominal/test_theils_u.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+@pytest.mark.parametrize("bias_correction", [False, True])
+@pytest.mark.parametrize("nan_strategy, nan_replace_value", [("replace", 0.0), ("drop", None)])
+class TestTheilsU(MetricTester):
+    atol = 1e-5
+
+    @pytest.mark.parametrize("ddp", [False, True])
+    @pytest.mark.parametrize("dist_sync_on_step", [False, True])
+    def test_theils_u(self, ddp, dist_sync_on_step, preds, target, bias_correction, nan_strategy, nan_replace_value):
+        metric_args = {
+            "bias_correction": bias_correction,
```
Comment (stancld) [inline tests/unittests/nominal/test_theils_u.py:111]: Just drop `bias_correction` from metric args as the argument is not used for this metric.

**Comment by stancld** [comment]
Also, can we add a link for this metric to `docs/source/links.rst`? -- https://en.wikipedia.org/wiki/Uncertainty_coefficient

**Comment by shenoynikhil** [comment]
@stancld @SkafteNicki I was wondering why in each of these (`crammers_v`, `pearson`, `theils_u`), we have the exact same update function. Would it make sense to have a common one so that it can be put into `utils.py`. I am referring to the following functions,
- `_pearsons_contingency_coefficient_update`
- `_cramers_v_update`
- `_pearsons_contingency_coefficient_update`
